### DISCUSSION TITLE
async hooks

### DISCUSSION THREAD

**Discussion by xurc** [mojolicious/mojo#1777] [Q&A] [answered]
Hello.

Is there any way to call to declare an asynchronous handler for "under" or before_dispatch / around_action hooks? Something like

`$app->hook(around_action => async sub { await ...  });`
`$route->under(async sub { await ...  });`

For example, if I need to perform some checks or fill a stash from the database and I want to use async / await in these operations.

How can wait non-blocking calls inside hooks or "under" in such a way that the execution is not transferred to the controller/action until all preliminary async operations have been completed? I also want to be able to cancel dispatch chain in such operations.

Is it possible?

**Comment by kraih** [comment, marked as answer]
I'm afraid it's not that easy yet. The problem is that `async/await` are only an optional dependency, so making the router work with it consistently would be very complicated. And backwards compatibility has to be considered as well.

**Reply by doojonio** [reply]
Is that possible to make it async with `continue` method in `Mojolicious::Controller`? Like here https://docs.mojolicious.org/Mojolicious/Guides/Routing#Under. In this case I think we can create `Mojo::Promise` object and in `then` callback call `continue` method

**Reply by xurc** [reply]
Yes, it works, thanks a lot for the tip.

```
            my $parent = $route->under(
                sub {
                   #
                   # Some code skipped
                   # 
                    my @promises;
                    for my $filler ( @{$fillers} ) {
                        # 'fill' declared as async
                        push @promises, $c->stashfiller->$filler->fill($c);
                    }

                    Mojo::Promise->all_settled(@promises)->then(
                        sub {
                            my @err;
                            for my $result (@_) {
                                push @err, $result->{value}[0] if $result->{value}[0];
                            }
                            return $c->respond_to( any => { $c->err->build( 409, \@err ) } ) if @err;
                            $c->continue;
                        }
                    )->wait;

                    return undef;
                }
            );
```
